### DISCUSSION TITLE
how is the service worker related to the project?

### DISCUSSION THREAD

**Discussion by wibed** [StopNGo/react-proto#6] [General]
from here 
```jsx
  if (IS_DEV) {
    if (hotReload != null) {
      app.get('/*', [...hotReload()])
    }
  } else {
    app.get('/sw.js', (_req, res) => {
      res.sendFile(path.join(SRC_DIR, 'sw.js'))
    })
  }
  ```

i derive it either is in mode `development` or sends the custom service worked within the response.
how does it fit together, given the variable `USE_SERVICE_WORKER` is false by default.

ps:

given the service worker is above only invoked if the file is accessed by url
and `USE_SERVICE_WORKER` is false by default
then how comes it into play?

```jsx
if (
  USE_SERVICE_WORKER &&
  String(process.env.NODE_ENV).trim() !== 'development'
) {
  const startServiceWorkerPromise = async (): Promise<void> => {
    const { startServiceWorker } = await import('./serviceWorker')
    startServiceWorker()
  }

  startServiceWorkerPromise().then(
    () => {},
    () => {}
  )
}
```

pps:
okey i guess the client conditionally registers a service worker by the dynamic url provided within the `getServiceWorker` function 
resulting in a request to the aforementioned url and resulting in the serviceWorker getting loaded, installed as outlined by the 
ServiceWorkerContainer API.

correct me if im wrong

**Comment by StopNGo** [comment]
Hi!
1. USE_SERVICE_WORKER is a constant for option. It is false by default just for purposes that SW is usually a last thing that you include into a project.
2. SW is included to the project if: USE_SERVICE_WORKER is true and it is a production mode , because SW during development mode could confuse you with caching.
3. Client conditionally registers a service worker via serviceWorker module with startServiceWorker function that registers sw.js file via navigator.serviceWorker API on window 'load' event.

But current version of react-proto (v1.4.1) has a bug and SW dose not start. It is because of my incorrect fixing TS promise error (in EventListener function should return void, not a Promise<void>).

This line will not start the registerSW function:
[serviceWorker.ts](https://github.com/StopNGo/react-proto/blob/main/src/serviceWorker.ts):
```js
window.addEventListener('load', () => registerSW)
```
so, it should be something like this:
```js
window.addEventListener('load', () => {
      (async () => {
        await registerSW()
      })().then(
        () => {}
      ).catch(er => console.log(er))
    })
```

I will fix it soon.
